### PR TITLE
_markNeedsClip() when clipper type has changed.

### PR DESCRIPTION
**Pull request by amirh**
A bug in _RenderCustomClip was compaeing the type of oldClipper to
itself instead to the type of newClipper.

This was the root cause for the crash #14937 worked around.
This also reverts the workaround introduced in #14937.

### CODE REVIEW

File: packages/flutter/lib/src/material/bottom_app_bar.dart [hunk 1]
```diff
@@ -155,9 +155,6 @@ class _BottomAppBarClipper extends CustomClipper<Path> {
 
   @override
   bool shouldReclip(CustomClipper<Path> oldClipper) {
```
Comment (Hixie) [inline packages/flutter/lib/src/material/bottom_app_bar.dart]: you can make the argument type `_BottomAppBarClipper` here, I believe the superclass explicitly makes it covariant.

File: packages/flutter/lib/src/material/bottom_app_bar.dart [hunk 2]
```diff
@@ -155,9 +155,6 @@ class _BottomAppBarClipper extends CustomClipper<Path> {
...
-    if (oldClipper.runtimeType != _BottomAppBarClipper)
-      return true;
-
     final _BottomAppBarClipper typedOldClipper = oldClipper;
```
Comment (Hixie) [inline packages/flutter/lib/src/material/bottom_app_bar.dart]: that would also let you get rid of this line, and could indeed turn the whole method into a one-liner.

**Comment by Hixie** [comment]
LGTM

test?

File: packages/flutter/lib/src/material/bottom_app_bar.dart (hunk 2 above)
Comment (amirh) [reply packages/flutter/lib/src/material/bottom_app_bar.dart]: done

File: packages/flutter/lib/src/material/bottom_app_bar.dart (hunk 1 above)
Comment (amirh) [reply packages/flutter/lib/src/material/bottom_app_bar.dart]: done

**Comment by amirh** [comment]
'toggle has notch' in bottom_app_bar_test is covering the bug

**Pull request merged**